pkg/networkserver/downlink.go

```diff
@@ -184,83 +197,111 @@ func (ns *NetworkServer) generateDownlink(ctx context.Context, dev *ttnpb.EndDev
 			logger.Debug("Need downlink for confirmed uplink")
 		}
 	}
-	if !needsDownlink &&
-		len(cmdBuf) == 0 &&
-		len(dev.QueuedApplicationDownlinks) == 0 {
-		return nil, errNoDownlink
-	}
-
-	pld := &ttnpb.MACPayload{
-		FHDR: ttnpb.FHDR{
-			DevAddr: dev.Session.DevAddr,
-			FCtrl: ttnpb.FCtrl{
-				Ack: up != nil && up.Payload.MHDR.MType == ttnpb.MType_CONFIRMED_UP,
-				ADR: deviceUseADR(dev, ns.defaultMACSettings),
-			},
-			FCnt: dev.Session.LastNFCntDown + 1,
-		},
-	}
-	logger = logger.WithField("ack", pld.FHDR.FCtrl.Ack)
 
-	var appDown *ttnpb.ApplicationDownlink
-	if len(cmdBuf) <= fOptsCapacity && len(dev.QueuedApplicationDownlinks) > 0 {
-		down := dev.QueuedApplicationDownlinks[0]
+	var start int
+	for _, down := range dev.QueuedApplicationDownlinks {
 		if len(down.FRMPayload) > int(maxDownLen) {
-			logger.Warn("Application downlink present, but the payload is too long, inform Application Server")
-			go func() {
-				ok, err := ns.handleASUplink(ctx, dev.EndDeviceIdentifiers.ApplicationIdentifiers, &ttnpb.ApplicationUp{
+			logger.Debug("Drop downlink with payload exceeding band regulations")
+			appUps.base = append(appUps.base, &ttnpb.ApplicationUp{
+				EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
+				CorrelationIDs:       append(events.CorrelationIDsFromContext(ctx), down.CorrelationIDs...),
+				Up: &ttnpb.ApplicationUp_DownlinkFailed{
+					DownlinkFailed: &ttnpb.ApplicationDownlinkFailed{
+						ApplicationDownlink: *down,
+						Error:               *ttnpb.ErrorDetailsToProto(errApplicationDownlinkTooLong),
+					},
+				},
+			})
+			start++
+			continue
+		}
+		if down.FCnt <= dev.Session.LastNFCntDown && dev.MACState.LoRaWANVersion.Compare(ttnpb.MAC_V1_1) < 0 {
+			logger.Debug("Drop downlink with too low FCnt")
+			appUps.base = append(appUps.base, &ttnpb.ApplicationUp{
+				EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
+				CorrelationIDs:       events.CorrelationIDsFromContext(ctx),
+				Up: &ttnpb.ApplicationUp_DownlinkQueueInvalidated{
+					DownlinkQueueInvalidated: &ttnpb.ApplicationInvalidatedDownlinks{
+						Downlinks:    dev.QueuedApplicationDownlinks,
+						LastFCntDown: dev.Session.LastNFCntDown + 1,
+					},
+				},
+			})
+			start++
+			continue
+		}
+		switch dev.MACState.DeviceClass {
+		case ttnpb.CLASS_A:
+			if down.GetClassBC() != nil {
+				logger.Debug("Drop class B/C downlink for device in class A mode")
```

Comment:
All devices are class A devices, even if class B or C is available. If application downlink indicates `class_b_c`, it should only be scheduled as class B/C downlink; as class A downlink it should stay in the queue, not dropped or failed.

Reply:
So it should be skipped then?

Reply:
Not skipped in terms of that we go to the next downlink; downlink should be transmitted in-order.

We should skip the downlink slot in class A if the first queued downlink is for class B/C only.